source/common/protobuf/deterministic_hash.cc

```diff
@@ -0,0 +1,206 @@
+#if defined(ENVOY_ENABLE_FULL_PROTOS)
+#include "source/common/protobuf/deterministic_hash.h"
+
+#include "source/common/common/assert.h"
+#include "source/common/common/hash.h"
+
+namespace Envoy {
+namespace DeterministicProtoHash {
+namespace {
+#define REFLECTION_FOR_EACH(get_type, hash_type)                                                   \
+  if (field->is_repeated()) {                                                                      \
+    for (int i = 0; i < reflection->FieldSize(message, field); i++) {                              \
+      hash_type(reflection->GetRepeated##get_type(message, field, i));                             \
+    }                                                                                              \
+  } else {                                                                                         \
+    hash_type(reflection->Get##get_type(message, field));                                          \
+  }
+
+#define MAP_SORT_BY(get_type)                                                                      \
```

Comment:
can you pass in the variable name rather than capturing the free variable, ie MAP_SORT_BY(map, get_type) ?

Also the `map` variable at call-sites is really a vector. Call it that? I'm actually not sure we really need to macro-izer the call to std::sort. I think you could do this readably without macros by instead just making a template function that returns the comparator:
```
std::sort(map.begin(), map.end(), makeCompareFn<get_type>);
```

I think that will be more readaable at call-sits and only slight more verbose.

Reply:
I don't think you can template e.g. `makeCompareFn<int>` to `reflection->GetInt(...)` can you? Other than by explicitly writing a specialization for each of the possible values. If we want to go that way I might as well just manually fully expand the macro throughout.
(It is frustrating that the reflection API doesn't have a Get<type> template with specializations that would make this easy!)

I've now done it with a middle-ground template which to me has the worst readability of all possible options. Do you prefer this over the full expansion?

Reply:
I see how that got awkward fast :) Let me reflect on that (no pun intended).

Reply:
Restructured it without that template, it's less lines of code.

Then separated the sorting out into its own helper class, then made the comparison function also its own helper class so there's no std::function or even lambdas involved any more. Then got rid of the original helper class because the comparer is enough to make it more suitable to just be a function again. :)